modules/transaction-payment/src/lib.rs

```diff
@@ -825,31 +821,101 @@ where
 					<T as Config>::Currency::ensure_can_withdraw(who, fee, reason, new_free_balance).is_ok()
 				});
 		if native_is_enough {
-			return;
+			None
+		} else {
+			Some(fee.saturating_add(native_existential_deposit.saturating_sub(total_native)))
 		}
+	}
+
+	/// Determine the fee and surplus that should be withdraw from user. There are three kind call:
+	/// - TransactionPayment::with_fee_path: swap with dex
+	/// - TransactionPayment::with_fee_currency: swap with tx fee pool
+	/// - others call: first use native asset, if not enough use alternative, or else use default.
+	fn ensure_can_charge_fee_with_call(
+		who: &T::AccountId,
+		fee: PalletBalanceOf<T>,
+		reason: WithdrawReasons,
+		call: &CallOf<T>,
+	) -> Result<Balance, DispatchError> {
+		let custom_fee_surplus = T::CustomFeeSurplus::get().mul_ceil(fee);
+		let custom_fee_amount = fee.saturating_add(custom_fee_surplus);
+		let fee_surplus: Result<Balance, DispatchError> = match call.is_sub_type() {
+			Some(Call::with_fee_path { fee_swap_path, .. }) => {
+				let supply_currency_id = *fee_swap_path.first().expect("should match a non native asset");
+				T::DEX::swap_with_specific_path(
+					who,
+					fee_swap_path,
+					SwapLimit::ExactTarget(
+						T::MultiCurrency::free_balance(supply_currency_id, who),
+						custom_fee_amount,
+					),
+				)
+				.map(|_| custom_fee_surplus)
+			}
+			Some(Call::with_fee_currency { currency_id, .. }) => {
+				Self::swap_from_pool_or_dex(who, custom_fee_amount, *currency_id).map(|_| custom_fee_surplus)
+			}
+			Some(_) | None => Self::native_then_alternative_or_default(who, fee, reason),
+		};
+		fee_surplus
+	}
 
-		// make sure add extra gap to keep alive after swap.
-		let amount = fee.saturating_add(native_existential_deposit.saturating_sub(total_native));
-		// native is not enough, try swap native from fee pool to pay fee and gap.
-		Self::swap_native_asset(who, amount);
+	fn ensure_can_charge_fee(who: &T::AccountId, fee: PalletBalanceOf<T>, reason: WithdrawReasons) {
```

Comment:
is this used?

Reply:
`reserve_fee` and `charge_fee` in `TransactionPayment` trait used it without `call` parameter.
- `reserve_fee` is used by evm schedule precompile.
- `charge_fee` is used by renvm-bridge(as it's not used, mabye need cleanup too)

Reply:
delete `ensure_can_charge_fee` invoke in `reserve_fee`, does this effect evm schedule precompile? @zjb0807